substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,830 @@
+	/// method.
+	fn on_validator_add(who: &T::AccountId, self_stake: Option<Stake<BalanceOf<T>>>) {
+		let self_stake = self_stake.unwrap_or_default().active;
+
+		if !T::TargetList::contains(who) {
+			T::TargetList::on_insert(who.clone(), self_stake)
+				.expect("staker does not exist in the list as per check above; qed.");
+		} else {
+			// if the target already exists in the list, it means that the target has been idle
+			// and/or dangling.
+			debug_assert!(
+				T::Staking::status(who) == Ok(StakerStatus::Idle) ||
+					T::Staking::status(who).is_err()
+			);
+
+			let self_stake = Self::to_vote_extended(self_stake);
+			Self::update_target_score(who, StakeImbalance::Positive(self_stake));
+		}
+
+		log!(debug, "on_validator_add: {:?}. role: {:?}", who, T::Staking::status(who),);
+
+		// a validator is also a nominator.
+		Self::on_nominator_add(who, vec![])
+	}
+
+	/// Fired when a validator becomes idle (i.e. chilling).
+	///
+	/// While chilled, the target node remains in the target list.
+	///
+	/// While idling, the target node is not removed from the target list but its score is
+	/// updated.
+	fn on_validator_idle(who: &T::AccountId) {
+		let self_stake = Self::weight_of(Self::active_vote_of(who));
+		Self::update_target_score(who, StakeImbalance::Negative(self_stake.into()));
+
+		// validator is a nominator too.
+		Self::on_nominator_idle(who, vec![]);
+
+		log!(debug, "on_validator_idle: {:?}, decreased self-stake {}", who, self_stake);
+	}
+
+	/// Fired when someone removes their intention to validate and has been completely removed
+	/// from the staking state.
+	///
+	/// The node is removed from the target list IFF its score is 0.
+	fn on_validator_remove(who: &T::AccountId) {
+		log!(debug, "on_validator_remove: {:?}", who,);
+
+		// validator must be idle before removing completely.
+		match T::Staking::status(who) {
+			Ok(StakerStatus::Idle) => (), // proceed
+			Ok(StakerStatus::Validator) => Self::on_validator_idle(who),
+			Ok(StakerStatus::Nominator(_)) => {
+				defensive!("on_validator_remove called on a nominator, unexpected.");
+				return
+			},
+			Err(_) => {
+				defensive!("on_validator_remove called on a non-existing target.");
+				return
+			},
+		};
+
+		// remove from target list IIF score is zero.
+		if T::TargetList::get_score(who).unwrap_or_default().is_zero() {
+			T::TargetList::on_remove(who).expect("target exists as per the check above; qed.");
+		}
+	}
+
+	/// Fired when someone sets their intention to nominate.
+	///
+	/// Note: it is assumed that `who`'s ledger staking state is updated *before* this method is
+	/// called.
+	fn on_nominator_add(who: &T::AccountId, nominations: Vec<AccountIdOf<T>>) {
+		let nominator_vote = Self::weight_of(Self::active_vote_of(who));
+
+		// voter may exist in the list in case of re-enabling a chilled nominator;
+		if T::VoterList::contains(who) {
+			return
+		}
+
+		let _ = T::VoterList::on_insert(who.clone(), nominator_vote)
```

Comment:
It would be easier to just handle the error in case of a duplicate instead of doing an extra access above.